test/common/upstream/load_stats_reporter_test.cc

```diff
@@ -221,6 +222,134 @@ TEST_F(LoadStatsReporterTest, ExistingClusters) {
   response_timer_cb_();
 }
 
+HostSharedPtr makeTestHost(const std::string& hostname,
+                           const ::envoy::config::core::v3::Locality& locality) {
+  const auto host = std::make_shared<NiceMock<::Envoy::Upstream::MockHost>>();
+  ON_CALL(*host, hostname()).WillByDefault(::testing::ReturnRef(hostname));
+  ON_CALL(*host, locality()).WillByDefault(::testing::ReturnRef(locality));
+  return host;
+}
+
+// Validate that per-locality metrics are aggregated across hosts and included in the load report.
+TEST_F(LoadStatsReporterTest, UpstreamLocalityStats) {
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  expectSendMessage({});
+  createLoadStatsReporter();
+  time_system_.setMonotonicTime(std::chrono::microseconds(3));
+
+  // Set up some load metrics
+  NiceMock<MockClusterMockPrioritySet> cluster;
+  MockHostSet& host_set_ = *cluster.prioritySet().getMockHostSet(0);
+
+  ::envoy::config::core::v3::Locality locality0, locality1;
+  locality0.set_region("mars");
+  locality1.set_region("jupiter");
+  HostSharedPtr host0 = makeTestHost("host0", locality0), host1 = makeTestHost("host1", locality0),
+                host2 = makeTestHost("host2", locality1);
+  host_set_.hosts_per_locality_ = makeHostsPerLocality({{host0, host1}, {host2}});
+
+  host0->stats().rq_success_.inc();
+  host0->loadMetricStats().add("metric_a", 0.11111);
```

Comment:
this test is very long and hard to see the structure of, or what it's trying to do. It's got a lot repetition. Suggest making helper functions in the LoadStatsReporterTest class, or maybe local lambdas that capture the repeated stanzas in one line. For example, you could capture each of the stanzas around here with something like:
```
  void addStats(const HostSharedPtr& host, float a, float b, float c = 0, float d = 0) {
    host->stats().rq_success_.inc();
    host->loadMetricStats().add("metric_a", a);
    host->loadMetricStats().add("metric_b", b);
    if (c != 0) { host->loadMetricStats().add("metric_c", c); }
    ...
}
```
and then these stanzas reduce to one line each:
```
   addStats(host0, 0.11111, 1.0);
   addStats(host0, 0.3333, 0, 3.14159);
   ....
```
Similar factoring can be done in the stanzas below with the expected stats setup. Those helper functions might might make it self-documenting, or we can add some comments.

Reply:
Done, agreed, it looks a lot better now.